lib/internal/fs/streams.js

```diff
@@ -386,22 +386,71 @@ WriteStream.prototype.open = openWriteFs;
 
 WriteStream.prototype._construct = _construct;
 
+function writeAll(data, size, pos, cb, retries = 0) {
+  this[kFs].write(this.fd, data, 0, size, pos, (er, bytesWritten, buffer) => {
+    if (er?.code === 'EAGAIN') {
+      er = null;
+      bytesWritten = 0;
+    }
+
+    if (this.destroyed || er) {
+      return cb(er);
+    }
+
+    this.bytesWritten += bytesWritten;
+
+    retries = bytesWritten ? 0 : retries + 1;
+
+    if (retries > 5) {
+      return cb(new Error('writev failed'));
+    }
+
+    if (bytesWritten < size) {
+      writeAll(buffer.slice(bytesWritten), pos + bytesWritten, cb, retries);
+    } else {
+      cb();
+    }
+  });
+}
+
+function writevAll(chunks, size, pos, cb, retries = 0) {
+  this[kFs].writev(this.fd, chunks, this.pos, (er, bytesWritten, buffers) => {
+    if (er?.code === 'EAGAIN') {
+      er = null;
+      bytesWritten = 0;
+    }
+
+    if (this.destroyed || er) {
+      return cb(er);
+    }
+
+    this.bytesWritten += bytesWritten;
+
+    retries = bytesWritten ? 0 : retries + 1;
+
+    if (retries > 5) {
+      return cb(new Error('writev failed'));
+    }
+
+    if (bytesWritten < size) {
+      writevAll([Buffer.concat(buffers).slice(bytesWritten)], pos + bytesWritten, cb, retries);
```

Comment:
I would recommend being a bit more subtle here and avoid the massive slowdown. Worst case you are allocating 2x memory to pick one byte. I would recommend just concatenating the buffers that have not been completely written.

Reply:
> Maybe add a comment here.

What kind of comment?